wrappers/desktop.go

```diff
@@ -235,50 +235,71 @@ func updateDesktopDatabase(desktopFiles []string) error {
 	return nil
 }
 
-// AddSnapDesktopFiles puts in place the desktop files for the applications from the snap.
-func AddSnapDesktopFiles(s *snap.Info) (err error) {
-	var created []string
-	defer func() {
-		if err == nil {
-			return
-		}
-
-		for _, fn := range created {
-			os.Remove(fn)
-		}
-	}()
-
-	if err := os.MkdirAll(dirs.SnapDesktopFilesDir, 0755); err != nil {
-		return err
+func findDesktopFiles(rootDir string) ([]string, error) {
+	if !osutil.IsDirectory(rootDir) {
+		return nil, nil
 	}
+	desktopFiles, err := filepath.Glob(filepath.Join(rootDir, "*.desktop"))
+	if err != nil {
+		return nil, fmt.Errorf("cannot get desktop files from %v: %s", rootDir, err)
+	}
+	return desktopFiles, nil
+}
 
-	baseDir := s.MountDir()
-
-	desktopFiles, err := filepath.Glob(filepath.Join(baseDir, "meta", "gui", "*.desktop"))
+func deriveDesktopFilesContent(s *snap.Info) (map[string]osutil.FileState, error) {
+	rootDir := filepath.Join(s.MountDir(), "meta", "gui")
+	desktopFiles, err := findDesktopFiles(rootDir)
 	if err != nil {
-		return fmt.Errorf("cannot get desktop files for %v: %s", baseDir, err)
+		return nil, err
 	}
 
+	content := make(map[string]osutil.FileState)
 	for _, df := range desktopFiles {
-		content, err := ioutil.ReadFile(df)
+		base := filepath.Base(df)
+		fileContent, err := ioutil.ReadFile(df)
 		if err != nil {
-			return err
+			return nil, err
 		}
-
 		// FIXME: don't blindly use the snap desktop filename, mangle it
 		// but we can't just use the app name because a desktop file
 		// may call the same app with multiple parameters, e.g.
 		// --create-new, --open-existing etc
-		installedDesktopFileName := filepath.Join(dirs.SnapDesktopFilesDir, fmt.Sprintf("%s_%s", s.DesktopPrefix(), filepath.Base(df)))
-		content = sanitizeDesktopFile(s, installedDesktopFileName, content)
-		if err := osutil.AtomicWriteFile(installedDesktopFileName, content, 0755, 0); err != nil {
-			return err
+		base = fmt.Sprintf("%s_%s", s.DesktopPrefix(), base)
+		installedDesktopFileName := filepath.Join(dirs.SnapDesktopFilesDir, base)
+		fileContent = sanitizeDesktopFile(s, installedDesktopFileName, fileContent)
+		content[base] = &osutil.MemoryFileState{
+			Content: fileContent,
+			Mode:    0644,
 		}
-		created = append(created, installedDesktopFileName)
+	}
+	return content, nil
+}
+
+// EnsureSnapDesktopFiles puts in place the desktop files for the applications from the snap.
+//
+// It also removes desktop files from the applications of the old snap revision to ensure
+// that only new snap desktop files exist.
+func EnsureSnapDesktopFiles(s *snap.Info) error {
+	if s == nil {
+		panic("internal error: snap info cannot be nil")
```

Comment:
I don't think we should panic here but instead return an (internal) error. The last panic like this (https://github.com/snapcore/snapd/pull/13107) caused us some grief :) It should lead to refreshes failing for the snap that we run into this with a clear error in the change (I think)

Reply:
Updated, Thank you!